### PR TITLE
ref. #783 target set not writable in Object.assign

### PR DESCRIPTION
**Pull request by tuchida**
ref. #783

https://www.ecma-international.org/ecma-262/11.0/index.html#sec-object.assign
> Perform ? Set(to, nextKey, propValue, true).

https://www.ecma-international.org/ecma-262/11.0/index.html#sec-set-o-p-v-throw
> Let success be ? O.[[Set]](P, V, O).
> If success is false and Throw is true, throw a TypeError exception.

https://www.ecma-international.org/ecma-262/11.0/index.html#sec-ordinary-object-internal-methods-and-internal-slots-set-p-v-receiver
https://www.ecma-international.org/ecma-262/11.0/index.html#sec-ordinaryset
https://www.ecma-international.org/ecma-262/11.0/index.html#sec-ordinarysetwithowndescriptor
> If IsDataDescriptor(ownDesc) is true, then
> If ownDesc.[[Writable]] is false, return false.

### CODE REVIEW

**Comment by tuchida** [comment]
How do you check the property descriptor?

For example:
https://github.com/mozilla/rhino/blob/41e57c749027ff20b3977a3b3b25e1b7ad6d45df/src/org/mozilla/javascript/NativeObject.java#L479-L480
Do you need to use `ScriptRuntime.toBoolean` or `ScriptableObject.getProperty`?
```js
Object.defineProperty(o, 'a', {
  configurable: 1 // truthy
});
```
```js
Object.defineProperty(o, 'a', {
  __proto__: {
    configurable: true // prototype chain
  }
});
```

**Comment by tuchida** [comment]
Sorry, My misunderstanding.
https://github.com/mozilla/rhino/blob/41e57c749027ff20b3977a3b3b25e1b7ad6d45df/src/org/mozilla/javascript/ScriptableObject.java#L216-L218
I'll correct it.

File: src/org/mozilla/javascript/NativeObject.java [hunk 1]
```diff
@@ -567,6 +567,13 @@ public Object execIdCall(IdFunctionObject f, Context cx, Scriptable scope,
               Scriptable sourceObj = ScriptRuntime.toObject(cx, thisObj, args[i]);
               Object[] ids = sourceObj.getIds();
               for (Object key : ids) {
+                if (targetObj instanceof ScriptableObject) {
+                  ScriptableObject desc = ((ScriptableObject) targetObj).getOwnPropertyDescriptor(cx, key);
+                  if (desc != null && isDataDescriptor(desc) &&
+                      Boolean.FALSE.equals(desc.get("writable"))) {
```
Comment (gbrail) [inline src/org/mozilla/javascript/NativeObject.java:573]: Can you take a look at the other implementation that checks for something like this, which is in ScriptableObject.checkPropertyChange and make it as much like that as you can?
That code uses an "isFalse" method to test for false, which is likely a more consistent way to do this -- these things tend to be very picky for edge cases.

**Review by gbrail — CHANGES_REQUESTED** [review]
Just a minor thing. Thanks for looking at this!

File: src/org/mozilla/javascript/NativeObject.java (hunk 1 above)
Comment (tuchida) [reply src/org/mozilla/javascript/NativeObject.java]: Thanks.  I changed to use isFalse. 959147a

File: src/org/mozilla/javascript/NativeObject.java [hunk 2]
```diff
@@ -567,6 +567,12 @@ public Object execIdCall(IdFunctionObject f, Context cx, Scriptable scope,
               Scriptable sourceObj = ScriptRuntime.toObject(cx, thisObj, args[i]);
               Object[] ids = sourceObj.getIds();
               for (Object key : ids) {
+                if (targetObj instanceof ScriptableObject) {
+                  ScriptableObject desc = ((ScriptableObject) targetObj).getOwnPropertyDescriptor(cx, key);
+                  if (desc != null && isDataDescriptor(desc) && isFalse(desc.get("writable"))) {
+                      throw ScriptRuntime.typeError1("msg.change.value.with.writable.false", key);
```
Comment (gbrail) [inline src/org/mozilla/javascript/NativeObject.java:573]: Thanks! This looks great except that since I just merged your other PR this is now deprecated. Can you please update it?

**Review by gbrail — CHANGES_REQUESTED** [review]

File: src/org/mozilla/javascript/NativeObject.java (hunk 2 above)
Comment (tuchida) [reply src/org/mozilla/javascript/NativeObject.java]: I changed it to use `ScriptRuntime.typeErrorById` and squashed it.

**Comment by gbrail** [comment]
Great. Thanks!
